larpandoracontent/LArTrackShowerId/TrackShowerIdFeatureTool.cc

```diff
@@ -199,147 +194,567 @@ void LinearFitFeatureTool::CalculateVariablesSlidingLinearFit(const pandora::Clu
+        rmsSlidingLinearFit = -1.f;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDLinearFitFeatureTool::ReadSettings(const TiXmlHandle xmlHandle)
+{
     PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
-        "MinClusterCaloHits", m_minClusterCaloHits));
+        "SlidingLinearFitWindow", m_slidingLinearFitWindow));
 
     PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_FOUND, !=, XmlHelper::ReadValue(xmlHandle,
-        "NearbyClusterDistance", m_nearbyClusterDistance));
+        "SlidingLinearFitWindowLarge", m_slidingLinearFitWindowLarge));
 
     return STATUS_CODE_SUCCESS;
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-MipEnergyFeatureTool::MipEnergyFeatureTool() :
-    m_mipCorrectionPerHit(1.f)
+ThreeDVertexDistanceFeatureTool::ThreeDVertexDistanceFeatureTool() :
+    m_slidingLinearFitWindow(10000)
 {
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-void MipEnergyFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
-    const pandora::Cluster *const pCluster)
+void ThreeDVertexDistanceFeatureTool::Run(SupportVectorMachine::DoubleVector &featureVector, const Algorithm *const pAlgorithm,
+    const pandora::ParticleFlowObject *const pInputPfo)
 {
     if (PandoraContentApi::GetSettings(*pAlgorithm)->ShouldDisplayAlgorithmInfo())
         std::cout << "----> Running Algorithm Tool: " << this->GetInstanceName() << ", " << this->GetType() << std::endl;
 
-    float mipEnergy(0.f);
+    //find the neutrino vertex
+    const VertexList *pVertexList = nullptr;
+    (void) PandoraContentApi::GetCurrentList(*pAlgorithm, pVertexList);
+
+    if (pVertexList->empty() || (pVertexList->size() != 1) || (VERTEX_3D != pVertexList->front()->GetVertexType()))
+    {
+        throw StatusCodeException(STATUS_CODE_FAILURE);
+    }
 
-    for (const OrderedCaloHitList::value_type &layerIter : pCluster->GetOrderedCaloHitList())
+    const Vertex *const nuVertex(pVertexList->front());
+    //find the particle vertex
+   float vertexDistance(-1.f);
+    try
     {
-        for (const CaloHit *const pCaloHit : *layerIter.second)
-            mipEnergy += pCaloHit->GetMipEquivalentEnergy();
+        const Vertex *const pVertex = LArPfoHelper::GetVertex(pInputPfo);
+        const CartesianVector nuPosition(nuVertex->GetPosition()), pfoPosition(pVertex->GetPosition());
+        vertexDistance = std::sqrt((nuPosition - pfoPosition).GetMagnitudeSquared());
```

Comment:
Can just call `GetMagnitude()` to avoid extra sqrt - might look cleaner/more standard.